### DISCUSSION TITLE
如何正确使用plugins功能？

### DISCUSSION THREAD

**Discussion by fatinghenji** [NateScarlet/auto-derby#66] [Q&A] [answered]
## example_custom_training_score.py
下面是我自己的理解：
```python
class Training(single_mode.Training):
    def score(self, ctx: single_mode.Context) -> float:
        ret = super().score(ctx) #赋予权重
        ret += self.stamina * 0.3 #体力的计算值变为原来的1.3倍
        return ret
```
**我这样的理解是正确的吗？**
这样是更改了某一项的权重，如果对数值有类似“达标线”的要求的话，应该怎么调整代码呢？
比如我期望养成如下的马，应该怎么调整呢？
|速度|耐力|力量|毅力|智力|
|-----|-----|-----|-----|-----|
|800|600|800|300|300|
## example_custom_race_score.py
下面是我自己的理解：
```python
class Race(single_mode.Race):
    def score(self, ctx: single_mode.Context) -> float:
        ret = super().score(ctx)
        if self.name == "有馬記念": #当比赛名称中含有“有馬記念”
            ret += 10 #评分点＋10
        return ret
```
这个相对于好理解一些，有下面两个问题想咨询下：
- 可以减评分吗？
 ```python
class Race(single_mode.Race):
    def score(self, ctx: single_mode.Context) -> float:
        ret = super().score(ctx)
        if self.name == "OP": 
            ret -= 10
        return ret
 ```
- 可以设定多个优先级吗？
 ```python
class Race(single_mode.Race):
    def score(self, ctx: single_mode.Context) -> float:
        ret = super().score(ctx)
        if self.name == "GⅠ": 
            ret += 20
        if self.name == "GⅡ": 
            ret += 10
        return ret
 ```
## example_custom_context.py
这个我基本上没看明白工作原理是什么，~我好笨~，特别是
```python
class Context(single_mode.Context):
    def next_turn(self) -> None:
        super().next_turn()
        auto_derby.config.pause_if_race_order_gt = {
            1: 5,
            2: 3,
            3: 2,
            4: 1,
        }[self.date[0]]
```
里面的
```
1: 5,
2: 3,
3: 2,
4: 1,
```
应该怎么理解呢？

**Comment by NateScarlet** [comment, marked as answer]
## example_custom_training_score.py

`super().score(ctx) ` 是调用父类的 `score` 方法实现，这里是获取插件未更改前的评分 或者你不调用也可以 那就得纯靠你自己算评分了

> 比如我期望养成如下的马，应该怎么调整呢？

按当前属性插值 如果没达到标准线就给更高的分，当然如果你把一个训练的分数给太高可能导致总分下降

比如速度我的实现是

https://github.com/NateScarlet/auto-derby/blob/a5a1177d11addbb902af6488d2d94b94bd06cb45/auto_derby/single_mode/training.py#L186-L190

当前0速每点速度给2分 当前300速每点速度给1分 当前600分每点速度给0.8 中间是插值
`_training_single_score` 和 `mathtools.integrate` 的实现是一样的 之后删掉只用 mathtools

## example_custom_race_score.py

只有想不到没有做不到

不过 name 是比赛名称 永远不会是 `OP` `GⅠ` 如果你要判断比赛级别用 `self.grade == Race.GRADE_G1`

## example_custom_context.py

`next_turn` 会在每次一个新的回合调用

`self.date` 是当前日期，`self.date[0]` 是年份 

这个就是第一年跑第五名不暂停 第二年跑第三名不暂停 URA跑第一名不暂停

**Reply by NateScarlet** [reply]
而且你这注释实际上就是把代码用中文重复了一遍 这种注释正常都是要去掉的

**Reply by fatinghenji** [reply]
确实是这样，但是对我这种小白来说还是会方便一点。
> 按回合数插值 如果没达到标准线就给更高的分

`回合数插值`是指的是育成回合吗？这个参数是在哪里？
>当前0速每点速度给2分 当前300速每点速度给1分 当前600分每点速度给0.8 中间是插值

`中间是插值`怎么理解？是我断句断错了吗？
> 不过 name 是比赛名称 永远不会是 OP GⅠ 如果你要判断比赛级别用 self.grade == Race.GRADE_G1

那么代码应该修改成这个样子吗？
```python
class Race(single_mode.Race):
   def score(self, ctx: single_mode.Context) -> float:
       ret = super().score(ctx)
       if self.grade == Race.GRADE_G1: 
           ret += 20
       if self.grade == Race.GRADE_G2: 
           ret += 10
       if self.grade == Race.GRADE_OP: 
           ret -= 10
       return ret
```
以后会考虑图形化生成比赛的插件吗？类似于这样
![image](https://user-images.githubusercontent.com/37131177/122576654-7ad56d00-d084-11eb-909d-b966af647fb0.png)
勾选想跑的比赛之后，自动形成plugins的插件，在养成的时候自动跑。

**Reply by NateScarlet** [reply]
`回合数插值` 是 self.turn_count() 我记错了训练没用回合数用的当前属性

`中间是插值` 是指 0 到 300 300 到 600 中间， 你不用把每个值都指定一个分数

> 那么代码应该修改成这个样子吗？

这样写可以实现效果，但是写得很啰嗦。 如果重复比较同一个变量应该直接用字典查找

https://github.com/NateScarlet/auto-derby/blob/a5a1177d11addbb902af6488d2d94b94bd06cb45/auto_derby/single_mode/race.py#L528-L537

> 以后会考虑图形化生成比赛的插件吗？

图形界面我不是很需求 插件完全有能力实现这个 比如用 Qt
python 写图形界面的教程应该不少
